zrml/prediction-markets/src/tests.rs

```diff
@@ -1476,6 +1476,250 @@ fn on_market_close_successfully_auto_closes_multiple_markets_after_stall() {
+        assert_eq!(Swaps::pool(0).unwrap().pool_status, PoolStatus::Initialized);
+        assert_eq!(Swaps::pool(1).unwrap().pool_status, PoolStatus::Initialized);
+    });
+}
+
+#[test]
+fn on_market_close_market_status_manager_exceeds_max_recovery_time_frames_after_stall() {
+    // We check that `on_market_close` works correctly even if a block takes much longer than 12sec
+    // to be produced and multiple markets are involved.
+    ExtBuilder::default().build().execute_with(|| {
+        // Mock last time frame to prevent it from defaulting.
+        LastTimeFrame::<Runtime>::set(Some(0));
+
+        let end: Moment = (5 * MILLISECS_PER_BLOCK).into();
+        let category_count = 3;
+        assert_ok!(PredictionMarkets::create_cpmm_market_and_deploy_assets(
+            RuntimeOrigin::signed(ALICE),
+            Asset::Ztg,
+            Perbill::zero(),
+            ALICE,
+            MarketPeriod::Timestamp(0..end),
+            get_deadlines(),
+            gen_metadata(50),
+            MarketType::Categorical(category_count),
+            Some(MarketDisputeMechanism::SimpleDisputes),
+            0,
+            LIQUIDITY,
+            vec![<Runtime as zrml_swaps::Config>::MinWeight::get(); category_count.into()],
+        ));
+        assert_ok!(PredictionMarkets::create_cpmm_market_and_deploy_assets(
+            RuntimeOrigin::signed(ALICE),
+            Asset::Ztg,
+            Perbill::zero(),
+            ALICE,
+            MarketPeriod::Timestamp(0..end),
+            get_deadlines(),
+            gen_metadata(50),
+            MarketType::Categorical(category_count),
+            Some(MarketDisputeMechanism::SimpleDisputes),
+            0,
+            LIQUIDITY,
+            vec![<Runtime as zrml_swaps::Config>::MinWeight::get(); category_count.into()],
+        ));
+
+        // This block takes much longer than 12sec, but markets and pools still close correctly.
+        set_timestamp_for_on_initialize(
+            end + (crate::MAX_RECOVERY_TIME_FRAMES + 1) * MILLISECS_PER_BLOCK as u64,
+        );
+        run_to_block(2); // Trigger `on_initialize`; must be at least block #2!
+
+        System::assert_last_event(
+            Event::RecoveryLimitReached { last_time_frame: 0, limit_time_frame: 6 }.into(),
+        );
+
+        // still active, not closed, because recovery limit reached
+        let market_after_close = MarketCommons::market(&0).unwrap();
+        assert_eq!(market_after_close.status, MarketStatus::Active);
+        let pool_after_close = Swaps::pool(0).unwrap();
+        assert_eq!(pool_after_close.pool_status, PoolStatus::Active);
+
+        let market_after_close = MarketCommons::market(&1).unwrap();
+        assert_eq!(market_after_close.status, MarketStatus::Active);
+        let pool_after_close = Swaps::pool(1).unwrap();
+        assert_eq!(pool_after_close.pool_status, PoolStatus::Active);
+    });
+}
+
```

Comment:
Can we add tests here that verify that the extrinsic fails if the markets _shouldn't_ open or close?

Reply:
7d4fe8f5a33c9e10cae082a06324de172cca9cb8